irmin: expose Tree.is_empty

CraigFe opened the pull request.

samoht commented:
LGTM but if we expose this we should make sure the result is consistent with the ones used with `repr` ; can we have tests for this?

CraigFe commented:
I agree it's worth having tests for this, even if the edge cases are a bit tedious to construct. Will do that now & we can look into fuzz-testing this API at some later point.

CraigFe commented:
I've changed the implementation of `is_empty` to pre-compute the hash of the empty tree, meaning it now
returns `bool` rather than `bool Lwt.t`, and also added some tests for equivalence with `Type.equal tree_t Tree.empty`.

Ngoguey42 commented on src/irmin/tree.ml:
```diff
@@ -663,18 +663,17 @@ module Make (P : Private.S) = struct
-              let res = is_empty_after_updates v um in
-              Lwt.return_ok res
-          | _ -> (
-              to_value t >|= function
```
I'm not sure it matters but `to_value` has the advantage of testing `cached_value`

Ngoguey42 commented on src/irmin/tree.ml:
```diff
@@ -663,18 +663,17 @@ module Make (P : Private.S) = struct
-              let res = is_empty_after_updates v um in
-              Lwt.return_ok res
-          | _ -> (
-              to_value t >|= function
```
I'm not sure it matters but `to_value` has the advantage of testing `cached_value`

CraigFe replied:
I'm not sure either, but for consistency with `cached_map` I've added a call to `cached_value`.

There must be a better way to handle this caching; will think about that in a later refactor.
